### PR TITLE
ticdc: wait TiCDC cluster becomes healthy after resigning owner

### LINKED ISSUE #4664: Wait TiCDC cluster becomes healthy after resigning owner
Wait TiCDC cluster becomes healthy after resigning owner, otherwise there may be latency spike.

image

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Close #4664

### What is changed and how does it work?

Wait TiCDC cluster becomes healthy after resigning owner, otherwise there may be latency spike.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

| master branch | this PR |
| -- | -- |
| image | image |

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
None
```

### CODE REVIEW

File: pkg/manager/member/ticdc_upgrader.go [hunk 1]
```diff
@@ -109,11 +114,22 @@ func (u *ticdcUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulS
...
 
-		err = gracefulShutdownTiCDC(tc, u.deps.CDCControl, u.deps.PodControl, pod, ordinal, "Upgrade")
+		err = gracefulDrainTiCDC(tc, u.deps.CDCControl, u.deps.PodControl, pod, ordinal, "Upgrade")
 		if err != nil {
 			return err
 		}
-		klog.Infof("ticdcUpgrade.Upgrade: %s has graceful shutdown in cluster %s/%s", podName, tc.GetNamespace(), tc.GetName())
+		klog.Infof("ticdcUpgrade.Upgrade: %s graceful drain TiCDC complete in cluster %s/%s", podName, tc.GetNamespace(), tc.GetName())
+		nextOrd, hasNext := nextOrdinal(ordinal)
+		if hasNext {
```
Comment (KanShiori) [inline pkg/manager/member/ticdc_upgrader.go:123]: `podOrdinals` may not contain 0. How about use `i == len(podOrdinals)-1 `

Comment (KanShiori) [reply pkg/manager/member/ticdc_upgrader.go:123]: BTW, how about adding some comments to describe the reason

File: pkg/manager/member/ticdc_scaler.go [hunk 2]
```diff
@@ -162,6 +207,62 @@ func gracefulShutdownTiCDC(
...
+	crossUpgradeVersion := semver.MustParse(ticdcCrossUpgradeVersion)
+	// TiCDC supports graceful upgrade since v6.3.0.
+	if podVersion.Compare(crossUpgradeVersion) < 0 {
+		return false, nil
+	}
+	// TiCDC does not support graceful upgrade that cross two major versions.
+	// E.g., Upgrading from 6.3.0 to 8.0.0 is not supported.
+	return (podVersion.Major() == currentVersion.Major() ||
+		podVersion.Major() == currentVersion.Major()-1), nil
+}
```
Comment (KanShiori) [inline pkg/manager/member/ticdc_scaler.go:264]: how about enhancing the pkg `pkg/utils/cmpver` to simplify this check?

File: pkg/manager/member/ticdc_scaler.go [hunk 3]
```diff
@@ -162,6 +206,82 @@ func gracefulShutdownTiCDC(
...
+		return false, controller.RequeueErrorf(
+			"ticdc.%s: cluster %s/%s fail to get TiCDC status, error: %v",
+			action, tc.GetNamespace(), tc.GetName(), err)
+	}
+	podVersion := status.Version
+	currentVersion := tc.Spec.TiCDC.Version
+	if currentVersion == nil {
+		currentVersion = &tc.Spec.Version
+	}
+
```
Comment (KanShiori) [inline pkg/manager/member/ticdc_scaler.go:239]: If the client use the `image` field, the version is empty. How about adding a field `TiCDCVersion()` to parse version, like `TiFlashVersion`.
```go
// TiFlashVersion returns the image version used by TiFlash.
//
// If TiFlash isn't specified, return empty string.
func (tc *TidbCluster) TiFlashVersion() string {
	if tc.Spec.TiFlash == nil {
		return ""
	}

	image := tc.TiFlashImage()
	colonIdx := strings.LastIndexByte(image, ':')
	if colonIdx >= 0 {
		return image[colonIdx+1:]
	}

	return "latest"
}
```

In the future, we just need modify this function when we remove the `image` filed.

**Review by KanShiori — APPROVED** [review]

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**